tpp-run/MLIRBench.cpp

```diff
@@ -222,6 +226,13 @@ void MLIRBench::printVector(Value vector) {
 }
 
 LogicalResult MLIRBench::printMemRef(mlir::Value memRef) {
+  OpBuilder::InsertionGuard guard(builder);
+
+  // Build print logic directly after the kernel return value
+  // before the output gets potentially deallocated.
+  if (isa<func::CallOp>(memRef.getDefiningOp()))
```

Comment:
While this works it's a little surprising after reading the function above and expecting the deallocation to happen just after the call, unless there's a print, etc. It can cause confusion on developers, for example, assuming the deallocation immediately after could optimize things, or end up exposing undefined behaviour.

I'd keep the `printMemref` as it is and move the deallocation to some new `cleanup` method that is called at the end, either by hand on the main file or from the lowering call (that also adds a return).

Reply:
With the previous approach of passing results through an output argument it was possible to have a single reusable buffer and just clean it up afterwards.
But now if we use function return for kernel output then each kernel call allocates memory internally. Which means that we have plenty of allocation in the main benchmarking loop so we need to clean it up immediately anyway.

I probably could pick the insertion point for `printMemRef` in a bit nicer way though.

Reply:
I rearranged the printing logic a bit to make it cleaner and decouple it from the explicit assumption that there is any dealloc.